crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -51,18 +52,254 @@ use std::{ops::Range, slice::SliceIndex};
+    pub representative_entity: Entity,
+
+    /// The last instance index in this batch.
+    pub last_instance_index: u32,
+
+    /// The dynamic offset of the batch.
+    ///
+    /// Note that dynamic offsets are only used on platforms that don't support
+    /// storage buffers.
+    pub dynamic_offset: Option<NonMaxU32>,
+}
+
+/// Information about the unbatchable entities in a bin.
+pub struct UnbatchableBinnedEntities {
+    /// The entities.
+    pub entities: Vec<Entity>,
+
+    /// Each entity's dynamic offset.
+    ///
+    /// Note that dynamic offsets are only used on platforms that don't support
+    /// storage buffers.
+    ///
+    /// This value may have fewer entries than entities. If so, the remaining
+    /// entries are assumed to be `None`. We do this to speed up platforms that
+    /// don't need dynamic offsets.
+    pub dynamic_offsets: Vec<Option<NonMaxU32>>,
+}
+
+impl<BPI> BinnedRenderPhase<BPI>
+where
+    BPI: BinnedPhaseItem,
+{
+    /// Bins a new entity.
+    ///
+    /// `batchable` specifies whether the entity can be batched with other
+    /// entities of the same type.
+    pub fn add(&mut self, key: BPI::BinKey, entity: Entity, batchable: bool) {
+        if batchable {
+            match self.batchable_values.entry(key.clone()) {
+                Entry::Occupied(mut entry) => entry.get_mut().push(entity),
+                Entry::Vacant(entry) => {
+                    self.batchable_keys.push(key);
+                    entry.insert(vec![entity]);
+                }
+            }
+        } else {
+            match self.unbatchable_values.entry(key.clone()) {
+                Entry::Occupied(mut entry) => entry.get_mut().entities.push(entity),
+                Entry::Vacant(entry) => {
+                    self.unbatchable_keys.push(key);
+                    entry.insert(UnbatchableBinnedEntities {
+                        entities: vec![entity],
+                        dynamic_offsets: vec![],
+                    });
+                }
+            }
+        }
+    }
+
+    /// Executes the GPU commands needed to render all entities in this phase.
+    pub fn render<'w>(
+        &self,
+        render_pass: &mut TrackedRenderPass<'w>,
+        world: &'w World,
+        view: Entity,
+    ) {
+        let draw_functions = world.resource::<DrawFunctions<BPI>>();
+        let mut draw_functions = draw_functions.write();
+        draw_functions.prepare(world);
+
+        // Draw batchables. Track instance index. We assume that all instance
+        // indices are consecutive.
+        let mut first_instance_index = self.first_instance_index;
+        debug_assert_eq!(self.batchable_keys.len(), self.batches.len());
+
+        for (key, batch) in self.batchable_keys.iter().zip(self.batches.iter()) {
+            let binned_phase_item = BPI::new(
+                key.clone(),
+                batch.representative_entity,
+                first_instance_index..batch.last_instance_index,
```

Comment:
`last_instance_index` feels incorrectly-named. It denotes the end of the instance range for the batch (like batch_range.end did). It isn't really the last instance index, it's the next instance index. Not sure what best to call it though.

Reply:
I renamed `first_instance_index` to `instance_start_index` and `last_instance_index` to `instance_end_index`.